File: library/src/test/java/com/coolerfall/download/DownloadManagerTest.java
package com.coolerfall.download;

import okhttp3.mockwebserver.MockWebServer;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowLog;

import static org.junit.Assert.assertEquals;

/**
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
@RunWith(RobolectricTestRunner.class) @Config(sdk = 29, manifest = Config.NONE)
public class DownloadManagerTest {
  private MockWebServer mockWebServer;
  private DownloadManager downloadManager;
  private DownloadRequest request;

  @Before public void setUp() throws Exception {
    ShadowLog.stream = System.out;
    mockWebServer = new MockWebServer();
    downloadManager = new DownloadManager.Builder().context(
        RuntimeEnvironment.getApplication().getApplicationContext()).build();
    request = new DownloadRequest.Builder().url(mockWebServer.url("/").toString())
        .relativeFilePath("/shadow/download.apk")
        .build();
  }

  @After public void tearDown() throws Exception {
    downloadManager.release();
    mockWebServer.shutdown();
  }

  @Test public void testDownload() throws Exception {
    int downloadId = downloadManager.add(request);
    assertEquals(1, downloadId);
  }

  @Test public void testDuplicate() throws Exception {
    int downloadId = downloadManager.add(request);
    assertEquals(1, downloadId);
    int downloadId2 = downloadManager.add(request);
    assertEquals(-1, downloadId2);
  }
}

File: library/src/test/kotlin/com/coolerfall/download/DownloadManagerTest.kt
package com.coolerfall.download

import com.coolerfall.download.DownloadManager.Builder
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowLog

/**
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [29], manifest = Config.NONE)
class DownloadManagerTest {
  private var mockWebServer: MockWebServer? = null
  private var downloadManager: DownloadManager? = null
  private var request: DownloadRequest? = null
  @Before @Throws(Exception::class) fun setUp() {
    ShadowLog.stream = System.out
    mockWebServer = MockWebServer()
    downloadManager = Builder().context(
        RuntimeEnvironment.getApplication().applicationContext
    )
        .build()
    request = DownloadRequest.Builder()
        .url(
            mockWebServer!!.url("/")
                .toString()
        )
        .relativeFilepath("/shadow/download.apk")
        .build()
  }

  @After @Throws(Exception::class) fun tearDown() {
    downloadManager!!.release()
    mockWebServer!!.shutdown()
  }

  @Test @Throws(Exception::class) fun testDownload() {
    val downloadId = downloadManager!!.add(request!!)
    Assert.assertEquals(1, downloadId.toLong())
  }

  @Test @Throws(Exception::class) fun testDuplicate() {
    val downloadId = downloadManager!!.add(request!!)
    Assert.assertEquals(1, downloadId.toLong())
    val downloadId2 = downloadManager!!.add(request!!)
    Assert.assertEquals(-1, downloadId2.toLong())
  }
}